### PR TITLE
BUG: levy_stable _fitstart fixes

### LINKED ISSUE #14470: levy_stable._fitstart() incorrect parameter estimates for negative beta, negative loc, or alpha==1
My issue is about scipy.levy_stable._fitstart().

- A data sample and its mirror image (negative values) should return the same estimate of alpha, but do not.
- A data sample and its mirror image should return mirror image estimates of the location parameter but do not.
- A data sample with a negative location and a shifted sample with a slightly different negative location should return shifted estimates of the location parameter but do not.
- A data sample with alpha==1 should not try to subtract infinity from a finite number but the calculation of the location parameter (delta) is trying to do that when it calculates tan(pi/2).  (My test code below induces an estimate of alpha==1 to exercise the relevant code, but does not currently check the value of the location parameter (delta) because I'm not sure what value is intended in this circumstance but I'm pretty sure it shouldn't involve trying to subtract tan(pi/2) from a finite value.)

#### Reproducing code example:

```
import pytest

# Test scipy.stats.levy_stable._fitstart().

import numpy as np
from scipy.stats import levy_stable
from scipy.stats import rv_continuous as rvc

class TestLevy_stable:
    def test_neg_beta(self):
        # Confirm that _fitstart() gets same estimate of alpha for +beta and -beta.
        rvc.random_state = 2021
        NPOINTS = 1001
        ALPHA = 1.5
        BETA_POS = 1
        
        draw = levy_stable.rvs(alpha=ALPHA, beta=BETA_POS, size=NPOINTS)
        # Fit stable distribution to positive and negative mirror image data, examine alpha.
        alpha_pos, alpha_neg = [levy_stable._fitstart(data)[0] for data in [draw, -draw]]
        
        assert np.isclose(alpha_pos, alpha_neg)
        
    def test_neg_delta(self):
        # Confirm that _fitstart() returns symmetric +delta and -delta if beta flips.
        rvc.random_state = 2021
        NPOINTS = 1001
        ALPHA = 1.5
        BETA_POS = 1
        
        draw = levy_stable.rvs(alpha=ALPHA, beta=BETA_POS, size=NPOINTS)
        # Fit stable distribution to positive and negative mirror image data, examine location.
        delta_pos, delta_neg = [levy_stable._fitstart(data)[2] for data in [draw, -draw]]
        
        assert np.isclose(delta_pos, -delta_neg)
        
    def test_delta_shift(self):
        # Confirm that _fitstart() returns delta that slides up and down if data shifts.
        rvc.random_state = 2021
        NPOINTS = 1001
        ALPHA = 1.5
        BETA_POS = 1
        LOC = -2
        SHIFT = 1
        
        draw = levy_stable.rvs(alpha=ALPHA, beta=BETA_POS, loc=LOC, size=NPOINTS)
        # Fit stable distribution to data and shifted data, examine location parameter.
        delta0, delta_up = [levy_stable._fitstart(draw + x)[2] for x in [0, SHIFT]]
        
        assert np.isclose(delta0, delta_up-SHIFT)
        
    def test_alpha1(self):
        # Confirm that stable distribution can fit alpha==1 without blowing up.
        
        # Make a dataset whose quintiles lead to alpha==1.
        x5 = x25 = x50 = 5  # Percentiles 5, 10, 25 -> 5.
        x95 = 100 - x5  # Percentile 95.
        # Calculate value for percentile 75 that will interpolate alpha lookup to 1.
        #   We assume beta==1 so interpolation is between rows 4 and 5 of last column
        #   of _fitstart() alpha_table.
        x75 = x25 + (x95 - x5) / (4 + (1.150 - 1) / (1.150 - .973))
        #assert x5 <= x25 <= x50 <= x75 <= x95
        
        xx = np.array([x5, x25, x50, x75, x95])
        # np.percentile(xx, [5, 25, 50, 75, 95]) -> [5, 5, 5, ..., 95]
        
        # Place target percentiles either side.
        data = np.array([x5, x5,  # Percentiles 0, 10.
                         x25, x25,  # 20, 30.
                         x50, x50, x50,  # 40, 50, 60.
                         x75, x75,  # 70, 80.
                         x95, x95  # 90, 100.
                         ])
        #assert np.array_equal(np.percentile(data, [5, 25, 50, 75, 95]), xx)
        
        alpha, beta, scale, loc = levy_stable._fitstart(data)

        # Not sure what value of delta should be returned when alpha==1 so
        #  for now just verify that alpha==1.
        assert alpha == 1
        # Might want to check loc too--what value should it have when alpha==1?
        
        
        
        

#%%

if __name__ == "__main__":
    t = TestLevy_stable()
    t.test_neg_beta()
    t.test_neg_delta()
    t.test_delta_shift()
    t.test_alpha1()
```

#### pytest results:
```
% pytest
================================================== test session starts ===================================================
platform darwin -- Python 3.7.6, pytest-6.2.4, py-1.10.0, pluggy-0.13.1
rootdir: /Users/...
collected 4 items                                                                                                        

test_stats.py FFF.                                                                                                 [100%]

======================================================== FAILURES ========================================================
_____________________________________________ TestLevy_stable.test_neg_beta ______________________________________________

self = <test_stats.TestLevy_stable object at 0x7fc2f4c94510>

    def test_neg_beta(self):
        # Confirm that _fitstart() gets same estimate of alpha for +beta and -beta.
        rvc.random_state = 2021
        NPOINTS = 1001
        ALPHA = 1.5
        BETA_POS = 1
    
        draw = levy_stable.rvs(alpha=ALPHA, beta=BETA_POS, size=NPOINTS)
        # Fit stable distribution to positive and negative mirror image data, examine alpha.
        alpha_pos, alpha_neg = [levy_stable._fitstart(data)[0] for data in [draw, -draw]]
    
>       assert np.isclose(alpha_pos, alpha_neg)
E       assert False
E        +  where False = <function isclose at 0x7fc2f3ccb5f0>(1.5529548477386994, 1.5672455848079985)
E        +    where <function isclose at 0x7fc2f3ccb5f0> = np.isclose

test_stats.py:21: AssertionError
_____________________________________________ TestLevy_stable.test_neg_delta _____________________________________________

self = <test_stats.TestLevy_stable object at 0x7fc2f4c94a10>

    def test_neg_delta(self):
        # Confirm that _fitstart() returns symmetric +delta and -delta if beta flips.
        rvc.random_state = 2021
        NPOINTS = 1001
        ALPHA = 1.5
        BETA_POS = 1
    
        draw = levy_stable.rvs(alpha=ALPHA, beta=BETA_POS, size=NPOINTS)
        # Fit stable distribution to positive and negative mirror image data, examine location.
        delta_pos, delta_neg = [levy_stable._fitstart(data)[2] for data in [draw, -draw]]
    
>       assert np.isclose(delta_pos, -delta_neg)
E       assert False
E        +  where False = <function isclose at 0x7fc2f3ccb5f0>(2.220446049250313e-16, -1.0714006456656149)
E        +    where <function isclose at 0x7fc2f3ccb5f0> = np.isclose

test_stats.py:34: AssertionError
____________________________________________ TestLevy_stable.test_delta_shift ____________________________________________

self = <test_stats.TestLevy_stable object at 0x7fc2f4ce2250>

    def test_delta_shift(self):
        # Confirm that _fitstart() returns delta that slides up and down if data shifts.
        rvc.random_state = 2021
        NPOINTS = 1001
        ALPHA = 1.5
        BETA_POS = 1
        LOC = -2
        SHIFT = 1
    
        draw = levy_stable.rvs(alpha=ALPHA, beta=BETA_POS, loc=LOC, size=NPOINTS)
        # Fit stable distribution to data and shifted data, examine location parameter.
        delta0, delta_up = [levy_stable._fitstart(draw + x)[2] for x in [0, SHIFT]]
    
>       assert np.isclose(delta0, delta_up-SHIFT)
E       assert False
E        +  where False = <function isclose at 0x7fc2f3ccb5f0>(2.220446049250313e-16, (2.220446049250313e-16 - 1))
E        +    where <function isclose at 0x7fc2f3ccb5f0> = np.isclose

test_stats.py:49: AssertionError
================================================ short test summary info =================================================
FAILED test_stats.py::TestLevy_stable::test_neg_beta - assert False
FAILED test_stats.py::TestLevy_stable::test_neg_delta - assert False
FAILED test_stats.py::TestLevy_stable::test_delta_shift - assert False
============================================== 3 failed, 1 passed in 5.47s ===============================================
```

#### Scipy/Numpy/Python version information:

```
1.5.2 1.18.5 sys.version_info(major=3, minor=7, micro=6, releaselevel='final', serial=0)
```

### PR DESCRIPTION
**Pull request by lutefiskhotdish**
#### Reference issue

See #14470

#### What does this implement/fix?

Adds new tests for levy_stable._fitstart(), and fixes bugs so the tests now pass.

- A data sample and its mirror image (negative values) now return the same estimate of alpha.
- A data sample and its mirror image now return mirror image estimates of the location parameter (delta).
- A data sample with a negative location and a shifted sample with a slightly different negative location now return shifted estimates of the location parameter.

#### Additional information

There is still something flaky about the calculation of the location parameter (delta) when alpha==1, as it tries to calculate tan(pi/2) -> infinity.  In practice, it will be very rare that alpha==1. when estimated from data samples.

### CODE REVIEW

File: scipy/stats/tests/test_levy_stable_fitstart.py
Comment (tylerjereddy) [inline scipy/stats/tests/test_levy_stable_fitstart.py:21]: Here and elsewhere, it is likely to be far more helpful to use `assert_allclose()` from the NumPy testing infrastructure than it is to use a plain `assert` on the boolean return value of the plain `isclose()` function, because `assert_allclose()` is specifically designed for testing purposes and clearly reporting the number and magnitude of differences.

File: scipy/stats/tests/test_levy_stable_fitstart.py
Comment (tylerjereddy) [inline scipy/stats/tests/test_levy_stable_fitstart.py:28]: here and elsewhere, I'm not sure much is gained by defining the variables like this and then using them only once vs. just using the values directly in the keyword arguments

File: scipy/stats/tests/test_levy_stable_fitstart.py
Comment (tylerjereddy) [inline scipy/stats/tests/test_levy_stable_fitstart.py:91]: I don't think we normally add blocks like this--instead, leaving the work to `pytest` to collect and run the tests automatically

**Review by tylerjereddy — COMMENTED** [review]
I'll defer to a `stats` regular on the actual fix--I just added a few "mechanical" comments re: the tests.

File: scipy/stats/tests/test_levy_stable_fitstart.py
Comment (tylerjereddy) [inline scipy/stats/tests/test_levy_stable_fitstart.py:77]: looks like the CI is complaining about this one: `E   NameError: name 'fitstart_fix' is not defined`

**Comment by ragibson** [comment]
Note that there is an ongoing overhaul of the levy_stable code at https://github.com/scipy/scipy/pull/9523.

This fix is partially correct, but is handled in a somewhat incorrect way.

Note that McCulloch's paper notes how to handle negative values in their interpolations.
1. Table 1:
image
2. Table 2:
image
3. Table 3:
image
4. Table 4:
image
5. Table 5:
image
6. Table 6:
image
7. Table 7:
image

Scipy's code is simply missing the negative beta case in psi_1. Look at
```python
# table III - alpha = psi_1(nu_alpha, nu_beta)
# table IV - beta = psi_2(nu_alpha, nu_beta)
# Table V - nu_c = psi_3(alpha, beta)
# Table VII - nu_zeta = psi_5(alpha, beta)
```
https://github.com/scipy/scipy/blob/88d4d946a22b3bd933884ec236bfba9526809ec9/scipy/stats/_continuous_distns.py#L5092-L5099
Really, I think you should add `psi_1_1` similar to the existing definition of `psi_2_1` and then use it in place of `psi_1` at
https://github.com/scipy/scipy/blob/88d4d946a22b3bd933884ec236bfba9526809ec9/scipy/stats/_continuous_distns.py#L5111-L5116

----------------

I think you are also correct that the clipping of delta to [eps, infty) is incorrect here:

https://github.com/scipy/scipy/blob/88d4d946a22b3bd933884ec236bfba9526809ec9/scipy/stats/_continuous_distns.py#L5117-L5121
Of course, the location of the distribution can be negative and I see no reason why the code does this unless negative delta is handled elsewhere and I missed it while skimming it.

----------------

> There is still something flaky about the calculation of the location parameter (delta) when alpha==1, as it tries to calculate tan(pi/2) -> infinity.

This is an unfortunate reality of the distribution itself in most common parameterizations (e.g. Zolotarev's / Nolan's S0). The mode of the distribution inherently diverges to infinity when delta is finite and alpha is close to (but not exactly) 1. McCulloch's paper has a few bits where this issue is mentioned, e.g.
image
Note: "if alpha is insignificantly different from unity, the true value of delta could lie anywhere".

File: scipy/stats/_continuous_distns.py
Comment (ragibson) [inline scipy/stats/_continuous_distns.py:5112]: For consistency, I'd suggest handling this in the same way that the current code does by adding a `psi_1_1` function:
https://github.com/scipy/scipy/blob/affdc4d54142cdcfedcfa06e1185c2ed0441c36c/scipy/stats/_continuous_distns.py#L5092-L5099

McCulloch's paper has the negative `nu_beta` case, but it was somehow missed in scipy's implementation.
image

The current code essentially treats negative `nu_beta` as `nu_beta == 0.0` in the interpolation, which is incorrect.

File: scipy/stats/tests/test_levy_stable_fitstart.py
Comment (lutefiskhotdish) [reply scipy/stats/tests/test_levy_stable_fitstart.py]: "far more helpful to use `assert_allclose()`"

Thanks.  That would be sound advice for sure in a context that could involve multiple values.  In `_fitstart()` the parameter lookups are assumed to be scalars so there is arguably less gain from the extra output of `assert_allclose()`, and perhaps potential for incorrectly implying that multiple values are potentially involved.  Thoughts?  Would you still advice using `assert_allclose()`?

File: scipy/stats/tests/test_levy_stable_fitstart.py
Comment (lutefiskhotdish) [reply scipy/stats/tests/test_levy_stable_fitstart.py]: Fair point, thanks.

File: scipy/stats/tests/test_levy_stable_fitstart.py
Comment (lutefiskhotdish) [reply scipy/stats/tests/test_levy_stable_fitstart.py]: Yikes!  Clearly juggling too many new tools at once and taking the wrong shortcuts.

File: scipy/stats/_continuous_distns.py
Comment (lutefiskhotdish) [reply scipy/stats/_continuous_distns.py:5112]: "I'd suggest handling this in the same way that the current code does by adding a psi_1_1 function"

In terms of consistency, you're probably right.

Any appetite for a slightly more ambitious revision with new helper functions `_rectified()` and `_reflected()` that would result in something like this:

```
def _rectified(func):
    return lambda xval, yval: func(val if xval > 0 else -xval, yval)

def _reflected(func):
    rectified_func = _rectified(func)
    def lookup(xval, yval):
        rectified_result = rectified_func(val,  yval)
        return rectified_result if xval > 0 else -rectified_result
    return lookup

psi_1_1 = _rectified(interpolate.interp2d(nu_beta_range, nu_alpha_range, alpha_table))
psi_2_1 = _reflected(interpolate.interp2d(nu_beta_range, nu_alpha_range, beta_table))

phi_3_1 = _rectified(interpolate.interp2d(beta_range, alpha_range, nu_c_table))
phi_5_1 = _reflected(interpolate.interp2d(beta_range, alpha_range, nu_zeta_table))

```

Or is it better to go with the minimal fix--less risk of breaking something that currently works?

Comment (lutefiskhotdish) [reply scipy/stats/_continuous_distns.py:5112]: Thanks for the excerpts from McCulloch--I haven't tracked down a copy yet.  McCulloch's equation 4.6 for the delta estimator suggests that the adjustment to zeta should be applied when alpha is **not** 1.  This is exactly the opposite of what the code is doing (I show it here with spurious clipping removed):

`delta = zeta-beta*c*np.tan(np.pi*alpha/2.) if alpha == 1. else zeta`

So I think the `==` should be `!=`.  Is that right?

If the true value could lie anywhere when alpha is exactly 1, there is arguably no point in checking for this event which is vanishingly unlikely anyway.  So maybe the code should just be:

`delta = zeta-beta*c*np.tan(np.pi*alpha/2.)`

Thoughts?

Comment (ragibson) [reply scipy/stats/_continuous_distns.py:5112]: I think you're right, but I'll have to check more closely tonight. There are many different parameterizations of stable laws and the literature has historically been very inconsistent in this regard.

I know scipy's current code uses Nolan's S1 parameterization and much of the theoretical work in papers used to use his S0 parameterization, which requires a shift of exactly the form used in the delta estimate here. It's unlikely, but that *could* explain the difference in the code.

For instance, Nolan notes the S0 <-> S1 transformation requires the following shift
image
(Also note that the field never really converged on consistent notation. Above, gamma plays a role similar to McCulloch's "c" in your code).

**Comment by ragibson** [comment]
@lutefiskhotdish I looked into this further and here are my suggested changes to the fix itself. I'm trying to keep the changes as minimal as reasonably possible so as to minimize the merging difficulties into the `levy_stable` overhaul in https://github.com/scipy/scipy/pull/9523.
```diff
diff --git a/scipy/stats/_continuous_distns.py b/scipy/stats/_continuous_distns.py
index 7d779e22a..070738737 100644
--- a/scipy/stats/_continuous_distns.py
+++ b/scipy/stats/_continuous_distns.py
@@ -5090,6 +5090,7 @@ class levy_stable_gen(rv_continuous):
             [0, -0.061, -0.279, -0.659, -1.198]]
 
         psi_1 = interpolate.interp2d(nu_beta_range, nu_alpha_range, alpha_table, kind='linear')
+        psi_1_1 = lambda nu_beta, nu_alpha: psi_1(nu_beta, nu_alpha) if nu_beta > 0 else psi_1(-nu_beta, nu_alpha)
         psi_2 = interpolate.interp2d(nu_beta_range, nu_alpha_range, beta_table, kind='linear')
         psi_2_1 = lambda nu_beta, nu_alpha: psi_2(nu_beta, nu_alpha) if nu_beta > 0 else -psi_2(-nu_beta, nu_alpha)
 
@@ -5109,14 +5110,14 @@ class levy_stable_gen(rv_continuous):
         nu_beta = (p95 + p05 - 2*p50)/(p95 - p05)
 
         if nu_alpha >= 2.439:
-            alpha = np.clip(psi_1(abs(nu_beta), nu_alpha)[0], np.finfo(float).eps, 2.)
+            alpha = np.clip(psi_1_1(nu_beta, nu_alpha)[0], np.finfo(float).eps, 2.)
             beta = np.clip(psi_2_1(nu_beta, nu_alpha)[0], -1., 1.)
         else:
             alpha = 2.0
             beta = np.sign(nu_beta)
         c = (p75 - p25) / phi_3_1(beta, alpha)[0]
         zeta = p50 + c*phi_5_1(beta, alpha)[0]
-        delta = zeta-beta*c*np.tan(np.pi*alpha/2.) if alpha == 1. else zeta
+        delta = zeta-beta*c*np.tan(np.pi*alpha/2.) if alpha != 1. else zeta
 
         return (alpha, beta, delta, c)
```

As mentioned in https://github.com/scipy/scipy/pull/14476#discussion_r677159863, note that the previous code mixes the cases for the delta estimate.

@lutefiskhotdish With the above patch, your tests still pass
```python
$ python runtests.py -v -m full -t scipy/stats/tests/test_levy_stable_fitstart.py 
Build OK (0:06:09.884302 elapsed)
================================================ test session starts =================================================
platform linux -- Python 3.9.6, pytest-6.2.4, py-1.10.0, pluggy-0.13.1 -- /usr/bin/python3.9
cachedir: .pytest_cache
rootdir: /home/ryan/PycharmProjects/scipy_stable/scipy, configfile: pytest.ini
collected 3 items                                                                                                    

scipy/stats/tests/test_levy_stable_fitstart.py::TestLevy_stable::test_neg_beta PASSED                          [ 33%]
scipy/stats/tests/test_levy_stable_fitstart.py::TestLevy_stable::test_neg_delta PASSED                         [ 66%]
scipy/stats/tests/test_levy_stable_fitstart.py::TestLevy_stable::test_delta_shift PASSED                       [100%]

================================================= 3 passed in 0.19s ==================================================
```
and the estimates are much more accurate
```python
>>> from scipy.stats import levy_stable
>>> 
>>> SIZE = 5 * 10 ** 7
>>> rvs = levy_stable.rvs(alpha=1.75, beta=0.5, loc=2.0, scale=5.0, size=SIZE)
>>> print(levy_stable._fitstart(rvs))
(1.7533970685117124, 0.554005532981971, 2.058881157647261, 4.990065245947207)
>>> 
>>> rvs = levy_stable.rvs(alpha=1.75, beta=-0.5, loc=2.0, scale=5.0, size=SIZE)
>>> print(levy_stable._fitstart(rvs))
(1.7535113451206517, -0.5563912462386369, 1.937892365646908, 4.989910501051077)
>>> 
>>> rvs = levy_stable.rvs(alpha=1.75, beta=0.5, loc=-2.0, scale=5.0, size=SIZE)
>>> print(levy_stable._fitstart(rvs))
(1.7530215468936947, 0.5526121180489312, -1.9417311451832033, 4.989026437213537)
>>> 
>>> rvs = levy_stable.rvs(alpha=1.75, beta=-0.5, loc=-2.0, scale=5.0, size=SIZE)
>>> print(levy_stable._fitstart(rvs))
(1.7529422179586065, -0.5530799346384316, -2.05956703533164, 4.989271645373516)
```
These tests estimate loc ~ 0.9 in your code and scipy's current implementation.

-----

@lutefiskhotdish You should also move your tests near the existing tests for `levy_stable._fitstart()` rather than making a completely new file. Somewhere around here:
https://github.com/scipy/scipy/blob/affdc4d54142cdcfedcfa06e1185c2ed0441c36c/scipy/stats/tests/test_distributions.py#L3028-L3052
perhaps with more specific names like `test_fit_neg_beta()`, `test_fit_neg_delta()`, and `test_fit_delta_shift()`.

The tests in that location also set random seeds with `np.random.seed()` rather than setting `rv_continuous.random_state` directly, so you'll probably have to change that to match.

I also think you might want to use `numpy.testing.assert_almost_equal()` for comparing scalars even if `assert_allclose()` technically works here.

File: scipy/stats/_continuous_distns.py
Comment (ragibson) [reply scipy/stats/_continuous_distns.py:5112]: It looks like swapping `alpha == 1` to `alpha != 1` is all that's needed for the delta estimation issue. More details are in https://github.com/scipy/scipy/pull/14476#issuecomment-887853522.

File: scipy/stats/tests/test_levy_stable_fitstart.py
Comment (lutefiskhotdish) [reply scipy/stats/tests/test_levy_stable_fitstart.py]: I've moved the tests to a better neighbourhood in test_distributions.py around line 3052 and rewritten them more in the style of what was there already.  The tests now mostly use numpy testing functions, except for a few plain asserts involving inequalities--is there a better way to do those?

File: scipy/stats/tests/test_levy_stable_fitstart.py
Comment (lutefiskhotdish) [reply scipy/stats/tests/test_levy_stable_fitstart.py]: This code block is not included in the new tests now.

**Comment by lutefiskhotdish** [comment]
@ragibson Thanks.  I think I've taken your advice on board in the latest commits but let me know if I've missed anything or if you have additional suggestsions.

**Comment by ragibson** [comment]
@lutefiskhotdish This looks mostly good to me. The CI linter does complain about a few formatting items:
```
scipy/stats/_continuous_distns.py:5093:9: E731 do not assign a lambda expression, use a def
scipy/stats/_continuous_distns.py:5093:80: E501 line too long (114 > 79 characters)
scipy/stats/_continuous_distns.py:5113:80: E501 line too long (83 > 79 characters)
scipy/stats/tests/test_distributions.py:3052:1: W293 blank line contains whitespace
scipy/stats/tests/test_distributions.py:3064:1: W293 blank line contains whitespace
scipy/stats/tests/test_distributions.py:3075:1: W293 blank line contains whitespace
```
After this, it looks like all the CI tests will pass, so you might have to go ahead and format the legacy code in `_fitstart()` according to PEP8 (c.f. https://github.com/scipy/scipy/pull/9523/files/34b9f6f14f3fe6c74200c8074b311b704f31b99f#r670516701).

I wonder if the scipy maintainers would be okay with disabling the lint check for those few lines in `_continuous_distns.py` just to make it easier to resolve merge conflicts with https://github.com/scipy/scipy/pull/9523 (where we have to resolve these PEP8 violations anyway). I suspect the answer is "no", but I admit I don't know exactly how they would handle this situation.

**Comment by lutefiskhotdish** [comment]
@ragibson I think the two commits I've just done make the bug fixes PEP8 compliant.  Am I to understand I should fix PEP8 violations in legacy code before BUG fixes can be taken on board?

**Comment by ragibson** [comment]
> @ragibson I think the two commits I've just done make the bug fixes PEP8 compliant. Am I to understand I should fix PEP8 violations in legacy code before BUG fixes can be taken onboard?

@lutefiskhotdish Typically you try to get the CI tests to pass before a PR is merged. Scipy has a tool that checks the diff for PEP8 violations, so editing legacy code (that is not PEP8 compliant) means you inherit those formatting errors from the perspective of the CI tests.

I see a few remaining PEP8 violations from your edits:
```
scipy/stats/_continuous_distns.py:5094:40: W291 trailing whitespace
scipy/stats/_continuous_distns.py:5095:80: E501 line too long (89 > 79 characters)
scipy/stats/_continuous_distns.py:5116:59: W291 trailing whitespace
```
For reference, this is from the CI failure at https://dev.azure.com/scipy-org/SciPy/_build/results?buildId=13326&view=logs&j=fdd89725-f657-5341-7d77-3c5c44c4f5b6&t=325fc03e-e0b9-5339-adaf-099a49a14600 and you can run similar checks locally with e.g.
```bash
$ python runtests.py -v --pep8
```
which gives me the same three warnings. There are some details on setting up your local environment correctly for this, but if you're interested, see https://scipy.github.io/devdocs/dev/contributor/runtests.html and https://scipy.github.io/devdocs/dev/contributor/pep8.html.

**Comment by lutefiskhotdish** [comment]
@ragibson Showing successful checks now so that must be progress, and now I know how to follow the trail of obscure links to find failing lint output....

**Comment by mdhaber** [comment]
@ragibson Was this one close enough to try to resolve the merge conflicts?

**Comment by ragibson** [comment]
> @ragibson Was this one close enough to try to resolve the merge conflicts?

Yes, the edits to `fitstart()` in `scipy/stats/_continuous_distns.py` here should map directly to
https://github.com/scipy/scipy/blob/a57451972d537b8e18a2f3badb482b7dc52ac4ef/scipy/stats/_levy_stable/__init__.py#L488-L629
and the edits in `scipy/stats/tests/test_distributions.py` can still be placed right after `levy_stable`'s `test_fit()`
https://github.com/scipy/scipy/blob/7818647e1d0c7c5a29bb975919c43c4687c34af3/scipy/stats/tests/test_distributions.py#L3293-L3307

**Comment by mdhaber** [comment]
Hmmm. Many of the changes in this PR were already present. You can compare by going back to [this commit](https://github.com/scipy/scipy/pull/14476/commits/724f2d1e91b59b2d9c3cf17a6a786ade70edf16b) and comparing against `main`. Maybe PEP8 made you make some of these changes when you moved to a new file?

I can make fixes if need be, but I'll let @steppi merge this one if it looks good.

**Comment by steppi** [comment]
> I can make fixes if need be, but I'll let @steppi merge this one if it looks good.

Thanks. I should have time to take a look tomorrow.

**Review by steppi — APPROVED** [review]
This looks good to me. I pushed one commit to change the `xfail` reason for `test_fit_rvs`. I had also written the previous reason; it referred to this PR as a possible fix and was written before I looked into what this PR is actually doing. The test failure is an unrelated timeout. @ragibson, if you have no objections, I'll merge this tomorrow.

**Review by ragibson — APPROVED** [review]
Yes, this looks good to me. The main fixes were addressing the missing negative sign in psi_1 and removing the clipping of delta to [eps, inf) since that parameter can definitely be negative.

The extra tests seem reasonable as well if they're passing.

**Comment by mdhaber** [comment]
Thanks @lutefiskhotdish and reviewers!

**Pull request merged**
